n3fit/src/n3fit/model_gen.py

```diff
@@ -572,18 +572,14 @@ def pdfNN_layer_generator(
     else:
         sumrule_layer = lambda x: x
 
-    # Only these layers change from replica to replica:
-    preprocessing_factor_replicas = []
-    for i_replica, replica_seed in enumerate(seed):
-        preprocessing_factor_replicas.append(
-            Preprocessing(
-                flav_info=flav_info,
-                input_shape=(1,),
-                name=f"preprocessing_factor_{i_replica}",
-                seed=replica_seed + number_of_layers,
-                large_x=not subtract_one,
-            )
-        )
+    compute_preprocessing_factor = Preprocessing(
+        flav_info=flav_info,
+        input_shape=(1,),
+        name="preprocessing_factor",
+        seed=seed[0] + number_of_layers,
```

Comment:
This means that training replicas 1 to 10 and 2 to 10 won't give me the same initial results, right?

However, would 1 to 10 and 1 to 9 produce the exact same result from 1 to 9? I would like to know that.

Reply:
You mean `n3fit 1 -r 10` vs `n3fit 2 -r 10`, and `n3fit 1 -r 10` versus `n3fit 1 -r 9`? I'll test, I'm actually not sure how those arguments get fed into `pdfNN_layer_generator`.

Reply:
Running in parallel I mean. You normally start at replica 1 but it is not necessarily always like that.

The reason why I'm asking is that I think right now if you instantiate, say, 10 replicas (to fit them in parallel for instance) each of them will get a separate seed for the generation of the pseudo-data and, until now, a separate seed for each of the Neural Networks.
As long as the seeds were separated that meant the initial point could be reproducible.

Instead, now all replicas are being instantiated with the first seed, so obviously if you fit 1 to 10 you will get a result different from 2 to 10 (because the corresponding first seed is different).

But I am wondering whether 1 to 10 would be equivalent to 1 to 9 for the first 9 replicas.

Then, one thing to be done for when parallel fit becomes the standard is recover that reproducibility. For now reproducibility is limited to sequential CPU replicas. Reproducibility cannot be achieved in any easy way in GPU, but at least the initial state can be made to be the same.

Reply:
I checked `n3fit 1 -r 4`, `n3fit 2 -r 4`, and `n3fit 1 -r 5`, and there were no similar results anywhere. For the `MultiDense` layers I did make sure that initialization is identical to master, here I did not.

Perhaps it's easiest to use the solution from there: [MultiInitializer](https://github.com/NNPDF/nnpdf/blob/341e39d60fb22383bd6c63559141288190360e86/n3fit/src/n3fit/backends/keras_backend/multi_dense.py#L148)?

Reply:
If it works out of the box for the preprocessing, yes indeed.
But if it doesn't we can leave that for later.

Btw, since you're at it, check that with seed[0] + number_of_layers we are not being silly and using the same seed twice. That was something that worked well when every layer was using seed +=1 but not sure whether that's the case anymore.

Reply:
If it's ok to leave for later, I'll try it in the next PR #1905, to save some hassle of moving the code between branches.

Reply:
The seeds should still be ok, I haven't changed the seeding of the NN layers, neither in this PR nor in the next.